src/foam/nanos/approval/ApprovableAwareDAO.js

```diff
@@ -0,0 +1,383 @@
+        throw new RuntimeException("The only approver of " + modelName + " is the maker of this request!");
+      }
+
+      // makers cannot approve their own requests even if they are an approver for the model
+      // however they will receive an approvalRequest which they can only view and not approve or reject
+      // so that they can keep track of the status of their requests
+      sendSingleRequest(x, request, request.getInitiatingUser());
+      approverIds.remove(request.getInitiatingUser());
+
+      for ( int i = 0; i < approverIds.size(); i++ ) {
+        sendSingleRequest(getX(), request, approverIds.get(i));
+      }
+      `
+    },
+    {
+      name: 'put_',
+      javaCode: `
+      User user = (User) x.get("user");
+      Logger logger = (Logger) x.get("logger");
+
+      LifecycleAware lifecycleObj = (LifecycleAware) obj;
+      
+      if ( ! getIsEnabled() ){
+        if ( lifecycleObj.getLifecycleState() == LifecycleState.PENDING ){
+          lifecycleObj.setLifecycleState(LifecycleState.ACTIVE);
+        }
+        return super.put_(x,obj);
+      }
+
+      // system and admins override the approval process
+      if ( user != null && ( user.getId() == User.SYSTEM_USER_ID || user.getGroup().equals("admin") || user.getGroup().equals("system") ) ) {
+        if ( lifecycleObj.getLifecycleState() == LifecycleState.PENDING && user.getId() != User.SYSTEM_USER_ID ){
+          lifecycleObj.setLifecycleState(LifecycleState.ACTIVE);
+        } 
+        else if ( lifecycleObj.getLifecycleState() == LifecycleState.PENDING && user.getId() == User.SYSTEM_USER_ID ) {
+          // Adding log message in case this change breaks something unexpected
+          Object primaryKey = obj instanceof foam.core.Identifiable ? ((foam.core.Identifiable)obj).getPrimaryKey() : null;
+          logger.warning("SYSTEM UPDATE - Not automatically setting LifecycleState from PENDING to ACTIVE for " + obj.getClass().getSimpleName() + ": " + primaryKey);
+        }
+        return super.put_(x,obj);
+      }
+
+      DAO approvalRequestDAO = (DAO) getX().get("approvalRequestDAO");
+      DAO dao = (DAO) getX().get(getDaoKey());
+
+      ApprovableAware approvableAwareObj = (ApprovableAware) obj;
+      FObject currentObjectInDAO = (FObject) dao.find(approvableAwareObj.getApprovableKey());
+      
+      if ( obj instanceof LifecycleAware && ((LifecycleAware) obj).getLifecycleState() == LifecycleState.DELETED ) {
+        approvableAwareObj = (ApprovableAware) currentObjectInDAO;
+
+        List approvedObjRemoveRequests = ((ArraySink) approvalRequestDAO
+          .where(
+            foam.mlang.MLang.AND(
+              foam.mlang.MLang.EQ(ApprovalRequest.DAO_KEY, getDaoKey()),
+              foam.mlang.MLang.EQ(ApprovalRequest.OBJ_ID, approvableAwareObj.getApprovableKey()),
+              foam.mlang.MLang.EQ(RoleApprovalRequest.OPERATION, Operations.REMOVE),
```

Comment:
Can this system work with the existing non role based approvals?

Reply:
So either:
1. Disable the approvableAwareDAO and make the desired model ApprovableAware (short term)
2. Don't add in the approvableAwareDAO at all in the services jrl
3. Migrate the OPERATION, INITIATING USER and IS_FULFILLED properties to the approvalRequest model